### ISSUE TITLE
Anon-delegates break with optional params.

### ISSUE THREAD

**Issue by gisikw** [KSP-KOS/KOS#1762]
Repro case:
```
global foo is { parameter a, b is 0. }.
global bar is { parameter c. }.
print "Program Runs".
```

Error:
```
 works when all mandatory parameters come before all optional parameters.
__________________________________________
           VERBOSE DESCRIPTION
While most commands in kOS work anywhere you
put them, there are a few exceptions in which
a command is only meaningful in some limited
places of the code.  This is one of those cases.
All defaultable parameters must come at the
end of the list of all parameters for a
program or a function.
For example, this is legal:
   declare x, y is 0.
but this is not:
   declare x is 0, y.
because when x had a default then all other
parameters that came after it had to have one.
__________________________________________
More Information at:
http://ksp-kos.github.io/KOS_DOC/language/variables/index.html#DECLARE_PARAMETER
__________________________________________
```

The following examples _do_ work:
```
global foo is { parameter a, b is 0. }.
print "Program Runs".
```

```
function foo { parameter a, b is 0. }
function bar { parameter c. }
print "Program Runs".
```

**Comment by Dunbaratu** [comment]
The problem is that the exception is being thrown from the compiler in a check it does at line 2249 of Compiler.cs, which was written before anon delegates existed.  It does a kind of "dumb" walk through the parse tree, expecting 'parameter' statements to only exist at the level of a function statement.  It should be easy to fix.

**Comment by Dunbaratu** [comment]
Looking at the code I can see that the solution will be at this point of the code:
https://github.com/KSP-KOS/KOS/blob/develop/src/kOS.Safe/Compilation/KS/Compiler.cs#L2269
Right now it recurses into a nested call of ``HasParameterStmtNested()`` for all parts of speech, with the exception of a declare function statement nested inside.  The case of the anon delegates has to be added to that check.  an anon delegate is, at the outer level of the braces, a generic expression.  So it needs to have the following added as an "and" to the check:

``&& child.Token.Type != TokenType.expr``

I say this because I'm not near my development computer at the moment and it may be possible for somebody else to try this out and see if it fixes it.  I'm analysing the problem entirely by looking at the code on github.

**Comment by gisikw** [comment]
Yep! Should be resolved via #1763 (apologies for the double-commits; forgot I have vim set up to strip whitespace, which caused a lot of superficial line changes)
